### ISSUE TITLE
Problem with Project64's protect memory option

### ISSUE THREAD

**Issue by purplemarshmallow** [gonetz/GLideN64#764]
There's a problem with Project64's protect memory option. The emulator gets unstable if it is enabled. This happens only with GLideN64 and with no other gfx plugin so there might be a problem on GLideN64's side. Protect Memory in Project64 uses special exeption handling (?) maybe GLideN64 has problems with it.

Zelda OOT: I haven't managed to reproduce a crash but others have #458 

Doubutsu No Mori: With the default settings (protect memory enabled) the game crashes for me. Not sure if this always happened but I think it was working before

Jet force Gemini: The game crashes during the intro cutscene and at saveral points ingame. If I disable protect memory and use savestates I can get past the point where the emulator crashes.
I debugged this one and noticed if protect memory is enabled PJ64 gives GLideN64 the command to fill RDRAM but with completely wrong values. GLideN64 does this and the emulator crashes

**Comment by purplemarshmallow** [comment]
glN64 does not have this problem if it's a regression it might be possible to see what changed since then. What makes it harder is that support for Project64 was removed during development and later re-added

**Comment by LegendOfDragoon** [comment]
>Protect Memory in Project64 uses special exception handling (?) maybe GLideN64 has problems with it.

That's what it looked like when I debugged it. Something to do with GlideN64.

**Comment by Cetacea3691** [comment]
Is this really a protect memory issue because Extreme-G and Beetle Adventure Racing! also suffer from this and they don't use protect memory.
This issue is so random in Extreme-G and Beetle Adventure Racing! that I can't find a reliable way to reproduce it.

**Comment by purplemarshmallow** [comment]
Yes this is a protect memory issue. If there's a problem with Extreme-G and Beetle Adventure Racing then it's not related. What exactly happens in these games?

**Comment by Cetacea3691** [comment]
They also have randomly the Unknown memory action errors like in #458

**Comment by purplemarshmallow** [comment]
What happens if you use CPU interpreter? Are the errors gone?

**Comment by purplemarshmallow** [comment]
and have you tested after https://github.com/gonetz/GLideN64/commit/4e415802fa2e20d516da6133286f85fbf91a63e0 ?

**Comment by Cetacea3691** [comment]
If I remember it right yes, but I'll have to do some extensive testing to verify it again.
I haven't tested it yet after 4e415802fa2e20d516da6133286f85fbf91a63e0.
Guess I'll be back in a few hours because of the randomness of those issues :)

**Comment by Cetacea3691** [comment]
I'm back from some extensive testing with updated GLideN64 and Project64.
There don't seem to be any more random Unknown memory action errors in Extreme-G and Beetle Adventure Racing! :)
p.s. Sorry for the inconvenience.

**Comment by purplemarshmallow** [comment]
I think it's good if problems like this are carefully tested

**Comment by purplemarshmallow** [comment]
I think there is something going wrong on both sides PJ64 and GLideN64. PJ64's recompiler is not always 100% accurate with protect memory enabled. But GLideN64's gDPFillRDRAM function can write out of RDRAM. It is checked if we are within RDRAM but I guess this check is not accurate enough and it can be more accurate to make sure.

**Comment by LegendOfDragoon** [comment]
Does using 16 MB RDRAM prevent the crash by any chance?

**Comment by purplemarshmallow** [comment]
It crashes even earlier
Most likely because GLideN64 does not support 16MB

**Comment by LegendOfDragoon** [comment]
I'd be willing to debug this, if someone can help me get this compiling in zilmar spec.

**Comment by purplemarshmallow** [comment]
https://github.com/purplemarshmallow/GLideN64/tree/NO_UI
You can merge this commit and then compile without UI

**Comment by LegendOfDragoon** [comment]
Cool, thanks for the advice and link.

I'd like to test that other JFG issue with the config as well, to confirm whether what I discovered is the sole cause of the issue. So I'll eventually need to compile the UI. But for now I guess I don't need a UI :smile: .

**Comment by gonetz** [comment]
> Doubutsu No Mori: With the default settings (protect memory enabled) the game crashes for me.

The problem still exists.
I disabled all writes to RDRAM but FrameBufferList::fillRDRAM to track it down.
The situation looks like this.
At the very first Display List, the game clears depth and color buffers with FillRect command.
First, it clears the depth buffer. Address of the depth buffer in RDRAM is 0x00194900. FillRect calls fillRDRAM, which fills RDRAM from 0x194900 to 0x1ba100. These writes to RDRAM cause no issues.
Then the game clears a color buffer. Address of the color buffer in RDRAM is 0x00000400. FillRect calls fillRDRAM, which tries to fill RDRAM from 0x0400 to 0x25c00, but the very first write to address 0x0400 causes the exception.
The situation looks absolutely valid. Corresponding part of log looks like:
```
0x00154138: CMD=0xFF W0=0xFF10013F W1=0x80000400
gDPSetColorImage( G_IM_FMT_RGBA, G_IM_SIZ_16b, 320, 0x00000400 );
0x00154140: CMD=0xE3 W0=0xE3000A01 W1=0x00300000
gSPSetOtherMode_H( G_CYC_FILL | ) result: 00382c30
0x00154148: CMD=0xE2 W0=0xE200001C W1=0x00000000
gSPSetOtherMode_L(  rendermode) result: 00000000
0x00154150: CMD=0xF7 W0=0xF7000000 W1=0x00010001
gDPSetFillColor( 0x00010001 );
0x00154158: CMD=0xF6 W0=0xF64FC3BC W1=0x00000000
fillRDRAM 0x400 - 0x25c00
Crash

```
@project64 could you please check, what is going wrong here? Is it possible to avoid this crash without disabling 'protect memory' option?

**Comment by project64** [comment]
I do not know about the exact situation. The way protect memory works is if there is code in that executed in the 4k page, then it changes the pages memory protection to read only. So any write even valid will cause it to generate an exception. 

Now there is two possible problems, either project64 is not handling the exception or project64 does not understand the opcode that is causing the exception. 

If it is is cause gliden64 is trying to handle the exception, you could put in a special condition by checking if the address is rdram and what the memory page state is.

**Comment by gonetz** [comment]
> If it is is cause gliden64 is trying to handle the exception, you could put in a special condition by checking if the address is rdram and what the memory page state is.

Do you know a portable way to know the memory page state?
I can check it with [VirtualQuery ](https://docs.microsoft.com/windows/desktop/api/memoryapi/nf-memoryapi-virtualquery) function, but it is Windows only. Have you implemented memory protection for Project64 for Android?

**Comment by project64** [comment]
I am not sure there is a good way to see the page table permission on linux/android, but if it is writing to the rdram address space is probably enough of a check.

**Comment by gonetz** [comment]
Check that write address is within RDRAM address space is not enough in case of Windows version of Project64. As I explained for Doubutsu No Mori case, plugin writes at address 0x00000400 , which is obviously within RDRAM. This write causes exception. I tested RDRAM + 0x00000400 address with  VirtualQuery and it returned that Protect status is PAGE_READONLY, so the exception. I can use that check on Windows and do not write to READONLY memory (it will cure the exception), but the question remains: why that page is READONLY if the game allocates that memory area for color buffer?

**Comment by LegendOfDragoon** [comment]
> project64 does not understand the opcode that is causing the exception.

Quite a few instructions aren't supported. For instance with Shunyuan's graphics plugin, it will crash in Super Smash Bros if Protect Memory is enabled because your exception handler doesn't handle `mov word ptr [eax+ecx*2],si`. Angrylion's has the same issue, I just don't remember which instruction it used when it crashes.

**Comment by project64** [comment]
if it is cause Project64 does not handle a particular opcode then there is something I can fix there.
